### PR TITLE
Part I - Setup Terraform Client for Project Neptune and Init Activity Implementation

### PR DESCRIPTION
**Pull request by Aayyush**
- Remove terraform client dependency on unused feature allocator 
- Copy over async client for Neptune along with Command Builder 
- Expose VersionLoader to be used by Project Neptune 
- Implement Terraform Init activity

### CODE REVIEW

File: server/neptune/terraform/async_client.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	tfDownloader terraform.Downloader,
+	usePluginCache bool,
+) (*AsyncClient, error) {
+	version, err := getDefaultVersion(defaultVersionStr, defaultVersionFlagName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting default version")
+	}
+
+	loader := terraform.NewVersionLoader(&terraform.DefaultDownloader{}, tfDownloadURL)
+
```
Comment (samrabelachew) [inline server/neptune/terraform/async_client.go:46]: the `github.com/runatlantis/atlantis/server/core/terraform` pkg has a lot of things going on that I think we should eventually break down. Can you instead pull out what you need into neptune like the async client?

**Review by samrabelachew — COMMENTED** [review]
took a first pass, left some initial comments

File: server/neptune/workflows/internal/activities/main.go [hunk 2]
```diff
@@ -37,13 +52,46 @@ type Terraform struct {
...
+	}
+
+	tfClient, err := terraform.NewAsyncClient(
+		binDir,
+		cacheDir,
+		config.DefaultVersionStr,
+		config.DefaultVersionFlagName,
+		config.DownloadURL,
+		&legacy_tf.DefaultDownloader{},
+		true,
```
[the remark this answers is not in the document]
Comment (Aayyush) [reply server/neptune/workflows/internal/activities/main.go:78]: We currently hardcode this when instantiating the terraform client in the legacy worker: https://github.com/lyft/atlantis/blob/1d8476419cbe01b736c97007d33eb439095f5505/server/server.go#L410-L419 
We could make this configurable but I don't see much value in it since disabling the plugin cache would mean downloading plugins for every root whenever we need to run a terraform command, which would also significantly spike our disk usage imo. 

I think we could just set the `terraformPluginCacheDir` in the command builder and avoid enabling/disabling the plugin cache for neptune. What do you think?

File: server/neptune/workflows/internal/activities/main.go [hunk 3]
```diff
@@ -37,13 +52,46 @@ type Terraform struct {
...
+func NewTerraform(config neptune.TerraformConfig, dataDir string, scope tally.Scope, serverURL *url.URL) (*Terraform, error) {
+	binDir, err := mkSubDir(dataDir, BinDirName)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheDir, err := mkSubDir(dataDir, TerraformPluginCacheDirName)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (Aayyush) [reply server/neptune/workflows/internal/activities/main.go:64]: Nish originally had a comment around this. If we do all the setup in the calling server file, it'll eventually be a giant file (similar to our current server.go file) which would be very hard to follow. So, I have tried to instantiate the common dependencies in the calling server file and the rest of the terraform specific stuff here. Since the binDir and cacheDir is only used by terraform activities, I think it makes sense to instantiate it here

File: server/neptune/config.go [hunk 4]
```diff
@@ -21,15 +21,23 @@ type ServerConfig struct {
 	Port    int
 }
 
+type TerraformConfig struct {
+	DefaultVersionStr      string
+	DefaultVersionFlagName string
+	DownloadURL            string
+	LogFilters             valid.TerraformLogFilters
```
[the remark this answers is not in the document]
Comment (Aayyush) [reply server/neptune/config.go:28]: Hmm I'm curious if we will need global config in the temporal worker for any other use cases since we build the root configs in the gateway. If we don't, I don't see much value in adding global config dependency just to pass in the log filters imo. What do you think?

File: server/neptune/terraform/async_client.go (hunk 1 above)
Comment (Aayyush) [reply server/neptune/terraform/async_client.go:46]: I just reused the version loader as it is since we didn't have to tweak the implementation, unlike async client which was tightly coupled with our legacy code (i.e `projectContext` and `projectCommandOutputHandler`). I don't see value in duplicating the code for Neptune if it is going to be exactly the same. 

However, it makes sense to do so If we plan to keep them isolated entirely. @nishkrishnan what do you think?

File: server/neptune/terraform/async_client.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+			outCh <- helpers.Line{Err: err}
+		} else {
+			c.Logger.InfoContext(ctx, fmt.Sprintf("successfully ran %q in %q", cmd.String(), path))
+		}
+	}()
+
+	return outCh
+}
+
+func (c *AsyncClient) WriteOutput(stdReader io.ReadCloser, outCh chan helpers.Line, jobID string) {
```
Comment (nishkrishnan) [inline server/neptune/terraform/async_client.go:148]: Can we get rid of `helpers`? like can we keep this in this package?

File: server/neptune/workflows/internal/activities/terraform.go [hunk 6]
```diff
@@ -1,16 +1,75 @@
...
+	terraformInitVerb := []string{"init"}
+	terraformInitArgs := []string{"-input=false"}
+	finalArgs := common.DeDuplicateExtraArgs(terraformInitArgs, request.Step.ExtraArgs)
+	terraformInitCmd := append(terraformInitVerb, finalArgs...)
+
+	outCh := t.TerraformExecutor.RunCommandAsync(ctx, request.JobID, request.Path, terraformInitCmd, request.Envs, tfVersion)
+	var lines []string
+	for line := range outCh {
+		if line.Err != nil {
+			err = line.Err
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform.go:61]: should we wrap this? i bet the error message wouldn't be obvious

File: server/neptune/workflows/internal/activities/terraform.go [hunk 7]
```diff
@@ -1,16 +1,75 @@
...
+			break
+		}
+		lines = append(lines, line.Line)
+	}
+	output := strings.Join(lines, "\n")
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform.go:66]: [rant] I wish we didn't have to add this, it assumes an implementation of the terraformClient, ideally we should just be retrieving a fixed buffer from the process and new lines should be encoded.

File: server/neptune/workflows/internal/activities/terraform.go [hunk 8]
```diff
@@ -1,16 +1,75 @@
...
+
+	// sanitize output by stripping out any ansi characters.
+	output = ansi.Strip(output)
+	return TerraformInitResponse{
+		Output: fmt.Sprintf("%s\n", output),
+	}, nil
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform.go:72]: Should we just not do this like i mentioned in slack, maybe we can just leave out outputs until we have a reasonable way to generate summaries?

File: server/neptune/workflows/internal/activities/terraform.go [hunk 9]
```diff
@@ -1,16 +1,75 @@
...
+	if version != nil {
+		tfVersion = version
+	}
+
+	terraformInitVerb := []string{"init"}
+	terraformInitArgs := []string{"-input=false"}
+	finalArgs := common.DeDuplicateExtraArgs(terraformInitArgs, request.Step.ExtraArgs)
+	terraformInitCmd := append(terraformInitVerb, finalArgs...)
+
+	outCh := t.TerraformExecutor.RunCommandAsync(ctx, request.JobID, request.Path, terraformInitCmd, request.Envs, tfVersion)
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform.go:57]: where is this written to our log handler?

File: server/neptune/workflows/internal/terraform/job/step/init/runner.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+type initActivities interface {
+	TerraformInit(ctx context.Context, request activities.TerraformInitRequest) (activities.TerraformInitResponse, error)
+}
+
+type Runner struct {
+	Activity initActivities
+}
+
+func (r *Runner) Run(executionContext *job.ExecutionContext, localRoot *root.LocalRoot, step job.Step) (string, error) {
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/terraform/job/step/init/runner.go:21]: feels a bit overkill to keep it in its own package don't you think? Should we just keep this in the job package? (Also feels like it doesn't need to be a struct and can just be another function imo)

File: server/neptune/workflows/terraform.go [hunk 11]
```diff
@@ -16,8 +20,11 @@ type TerraformActivities struct {
 	activities.Terraform
 }
 
-func NewTerraformActivities(scope tally.Scope, serverURL *url.URL) (*TerraformActivities, error) {
-	terraformActivities := activities.NewTerraform(serverURL)
+func NewTerraformActivities(config neptune.TerraformConfig, dataDir string, scope tally.Scope, serverURL *url.URL, logger logging.Logger) (*TerraformActivities, error) {
```
Comment (nishkrishnan) [inline server/neptune/workflows/terraform.go:23]: ah i see why you moved TerraformConfig, but i would just move that here instead of keeping it in `neptune/config`

Comment (nishkrishnan) [reply server/neptune/workflows/terraform.go:23]: I don't think we need to pass scope and logger into our activities, for scope we have:
https://github.com/temporalio/samples-go/blob/main/metrics/workflow.go#L36

and for logging we have:
https://github.com/lyft/atlantis/blob/25d3a061e24f5658a70889d9eed73d1cb81bbc6b/server/neptune/workflows/internal/activities/config/logger/logger.go#L10-L11

**Review by nishkrishnan — COMMENTED** [review]
Took a first pass

File: server/neptune/config.go [hunk 12]
```diff
@@ -1,4 +1,4 @@
-package temporalworker
+package neptune
```
[the remark this answers is not in the document]
Comment (Aayyush) [reply server/neptune/config.go]: I originally moved it out to remove circular dependency but I guess it should be inside temporalworker. So, I've moved it to `temporalworker/config/config.go`

File: server/neptune/terraform/async_client.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+//
+//     Terraform v0.11.10
+//	   => 0.11.10
+var versionRegex = regexp.MustCompile("Terraform v(.*?)(\\s.*)?\n")
+
+func NewAsyncClient(
+	binDir string,
+	cacheDir string,
+	defaultVersionStr string,
+	defaultVersionFlagName string,
```
[the remark this answers is not in the document]
Comment (Aayyush) [reply server/neptune/terraform/async_client.go]: Not really. We only use it for the error message! So, I've removed it

File: server/neptune/workflows/internal/activities/terraform.go (hunk 7 above)
Comment (Aayyush) [reply server/neptune/workflows/internal/activities/terraform.go:59]: do you mean we should encode the newline character when writing logs to the output channel here: https://github.com/lyft/atlantis/blob/65eb072e577a443a8e517656d7b0a5ff7fde1460/server/neptune/terraform/async_client.go#L162-L165

File: server/neptune/workflows/internal/activities/terraform.go (hunk 9 above)
Comment (Aayyush) [reply server/neptune/workflows/internal/activities/terraform.go]: I have a follow up PR for writing logs to the log handler. It should be similar to what we already have in async_client: 
https://github.com/lyft/atlantis/blob/65eb072e577a443a8e517656d7b0a5ff7fde1460/server/core/terraform/async_client.go#L130-L134
unless we plan to do it differently!

File: server/neptune/workflows/internal/terraform/job/step/init/runner.go (hunk 10 above)
Comment (Aayyush) [reply server/neptune/workflows/internal/terraform/job/step/init/runner.go:21]: Do you meant just for init or the other steps too? Yeah I agree it looks like an overkill. I originally organized it like this assuming we'd have to do some deterministic work before orchestrating the activity.

File: server/neptune/workflows/terraform.go (hunk 11 above)
Comment (Aayyush) [reply server/neptune/workflows/terraform.go]: Gotcha, we'd still need to build the metric name similar to how we build scope right? Also, I have wrapped the logging methods in an interface so that we can mock it for tests without having to worry about workflow and activity contexts. What do you think?

File: server/neptune/terraform/async_client.go (hunk 1 above)
Comment (nishkrishnan) [reply server/neptune/terraform/async_client.go]: yeah fine to just move that stuff later

File: server/neptune/workflows/terraform.go (hunk 11 above)
Comment (nishkrishnan) [reply server/neptune/workflows/terraform.go]: Not really a fan of this, we went from not needing to inject our logger at all to injecting it.  The struct doesn't have any state that we need to maintain at all either.

I'd rather you see what other folks using temporal are doing first before needing to do this (ie. by asking in the OSS channel, or looking at samples).

File: server/neptune/terraform/async_client.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+// Line represents a line that was output from a terraform command.
+type Line struct {
+	// Line is the contents of the line (without the newline).
+	Line string
+	// Err is set if there was an error.
+	Err error
+}
+
+// Setting the buffer size to 10mb
+const BufioScannerBufferSize = 10 * 1024 * 1024
```
Comment (nishkrishnan) [inline server/neptune/terraform/async_client.go:34]: nit: can this be private?

File: server/neptune/workflows/internal/activities/terraform.go [hunk 15]
```diff
@@ -1,16 +1,66 @@
...
+	// Resolve the tf version to be used for this operation
+	tfVersion, err := t.resolveVersion(request.TfVersion)
+	if err != nil {
+		return TerraformInitResponse{}, err
+	}
+
+	// Build tf command
+	terraformInitCmd := terraform.CommandArguments{
+		Command:     terraform.Init,
+		CommandArgs: []string{"-input=false"},
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform.go:46]: nit: This is an improvement, I'd probably also create a wrapper around the array itself, like

```
type Arg struct {
  key string
  value string
}
```

and have a method on it that builds out the string form, but its minor.

File: server/neptune/terraform/command.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+		return "init"
+	case Plan:
+		return "plan"
+	case Apply:
+		return "apply"
+	}
+	return ""
+}
+
+type CommandArguments struct {
```
Comment (nishkrishnan) [inline server/neptune/terraform/command.go:25]: do these all need to be exported?

File: server/neptune/terraform/command_builder.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/hashicorp/go-version"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/core/runtime/cache"
+)
+
+type CommandBuilder struct {
```
Comment (nishkrishnan) [inline server/neptune/terraform/command_builder.go:14]: does this need to be exported?

File: server/neptune/workflows/internal/activities/terraform.go [hunk 18]
```diff
@@ -1,16 +1,66 @@
...
+
+func (t *terraformActivities) TerraformInit(ctx context.Context, request TerraformInitRequest) (TerraformInitResponse, error) {
+	// Resolve the tf version to be used for this operation
+	tfVersion, err := t.resolveVersion(request.TfVersion)
+	if err != nil {
+		return TerraformInitResponse{}, err
+	}
+
+	// Build tf command
+	terraformInitCmd := terraform.CommandArguments{
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform.go:44]:
```suggestion
	cmd := terraform.CommandArguments{
```

we already know it's terraform init based on the function, keep var names concise

File: server/neptune/workflows/internal/activities/terraform.go [hunk 19]
```diff
@@ -1,16 +1,66 @@
...
+		Command:     terraform.Init,
+		CommandArgs: []string{"-input=false"},
+		ExtraArgs:   request.Step.ExtraArgs,
+	}.Build()
+
+	outCh := t.TerraformClient.RunCommand(ctx, request.JobID, request.Path, terraformInitCmd, request.Envs, tfVersion)
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform.go:50]: nit: same comment about naming here, either output or ch is fine

File: server/neptune/workflows/internal/terraform/job/step/init/runner.go (hunk 10 above)
Comment (nishkrishnan) [reply server/neptune/workflows/internal/terraform/job/step/init/runner.go:21]: doesn't seem like there's any thing else going on, let's just leave it in the job package

File: server/neptune/terraform/command.go (hunk 16 above)
Comment (Aayyush) [reply server/neptune/terraform/command.go]: We can export CommandArguments and keep everything else internal since it's used in the terraform activity.

**Review by nishkrishnan — APPROVED** [review]
Nice job, im assuming you've tested this locally?

File: server/neptune/workflows/internal/activities/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (samrabelachew) [reply server/neptune/workflows/internal/activities/main.go]: We can leave as is to unblock, it was more of a comment on how we are passing in a hardcoded boolean. Seems like we don't ever want the situation for it to be false, so if it is not a configuration, maybe we can get rid of that parameter and just make this the only supported behavior in the future.

File: server/neptune/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (samrabelachew) [reply server/neptune/config.go]: Currently LogFilters are supported as a global configuration, so I'd vote for not creating another separate location for setting them. In general, I think we want to [move towards using global config](https://github.com/lyft/atlantis/pull/275#discussion_r924966748) whenever possible to be our source of truth.

File: server/neptune/workflows/internal/terraform/job/init_step_runner.go [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+type InitStepRunner struct {
+	Activity initActivities
+}
+
+func (r *InitStepRunner) Run(executionContext *job.ExecutionContext, localRoot *root.LocalRoot, step job.Step) (string, error) {
+	var resp activities.TerraformInitResponse
+	err := workflow.ExecuteActivity(executionContext.Context, r.Activity.TerraformInit, activities.TerraformInitRequest{
+		Step:      step,
+		Envs:      executionContext.Envs,
+		TfVersion: executionContext.TfVersion,
```
Comment (samrabelachew) [inline server/neptune/workflows/internal/terraform/job/init_step_runner.go:26]: can you pass in the terraform version into the activity? I think it currently uses the local root, which doesn't assign that field yet. https://github.com/lyft/atlantis/blob/4da3ccf194c53ed14aaa26e9712c62aff4263d39/server/neptune/workflows/internal/terraform/job/job_runner.go#L42

**Comment by Aayyush** [comment]
> Nice job, im assuming you've tested this locally?

Yup, I had to pull in changes from Samra's other [PR](https://github.com/lyft/atlantis/pull/334) and it worked as expected!

File: server/neptune/gateway/event/push_event_handler.go [hunk 21]
```diff
@@ -104,6 +105,12 @@ func (p *PushHandler) handle(ctx context.Context, event Push) error {
 
 func (p *PushHandler) startWorkflow(ctx context.Context, event Push, rootCfg *valid.MergedProjectCfg) (client.WorkflowRun, error) {
 	options := client.StartWorkflowOptions{TaskQueue: workflows.DeployTaskQueue}
+
+	var tfVersion string
+	if rootCfg.TerraformVersion != nil {
+		tfVersion = rootCfg.TerraformVersion.String()
+	}
```
Comment (samrabelachew) [inline server/neptune/gateway/event/push_event_handler.go:112]: Should we fail if `tfVersion` is empty?

File: server/neptune/workflows/internal/activities/terraform.go [hunk 22]
```diff
@@ -93,6 +93,11 @@ func (t *terraformActivities) TerraformApply(ctx context.Context, request Terraf
 }
 
 func (t *terraformActivities) resolveVersion(v string) (*version.Version, error) {
+	// Use default version if comfigured version is empty
```
Comment (samrabelachew) [inline server/neptune/workflows/internal/activities/terraform.go]: `configured`

File: server/neptune/gateway/event/push_event_handler.go (hunk 21 above)
Comment (Aayyush) [reply server/neptune/gateway/event/push_event_handler.go:112]: this version is parsed from the atlantis.yaml file which does not enforce specifying the terraform versions for the root. So, if the version is not specified, we use the default tf version configured during startup. This is how atlantis currently handles terraform version afaik!

**Review by samrabelachew — APPROVED** [review]
